KAFKA-18220: Refactor AsyncConsumerMetrics to not extend KafkaConsumerMetrics

ShivsundarR opened the pull request:
*What*
https://issues.apache.org/jira/browse/KAFKA-18220

- Currently, `AsyncConsumerMetrics` extends `KafkaConsumerMetrics`, but
is being used both by `AsyncKafkaConsumer` and `ShareConsumerImpl`.

- `ShareConsumerImpl` only needs the async consumer metrics(the metrics
associated with the new consumer threading model).
- This needs to be fixed, we are unnecessarily having
`KafkaConsumerMetrics` as a parent class for `ShareConsumer` metrics.

Fix :
- In this PR, we have removed the dependancy of `AsyncConsumerMetrics`
on `KafkaConsumerMetrics` and made it an independent class which both
`AsyncKafkaConsumer` and `ShareConsumerImpl` will use.
- The "`asyncConsumerMetrics`" field represents the metrics associated
with the new consumer threading model (like application event queue
size, background queue size, etc).
- The "`kafkaConsumerMetrics`" and "`kafkaShareConsumerMetrics`" fields
denote the actual consumer metrics for `KafkaConsumer` and
`KafkaShareConsumer` respectively.

Reviewers: Andrew Schofield <aschofield@confluent.io>

ShivsundarR committed:
Refactor AsyncConsumerMetrics to not extend KafkaConsumerMetrics

ShivsundarR committed:
Modify test and update metric

AndrewJSchofield commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/metrics/AsyncConsumerMetricsTest.java:
```diff
@@ -47,21 +47,6 @@ public void tearDown() {
-            )
-        );
-
         Set<MetricName> expectedConsumerMetrics = Set.of(
```
I suggest renaming this to `expectedMetrics`. It is specifically NOT the expected consumer metrics :)

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:
```diff
@@ -2270,10 +2276,14 @@ public Metrics metricsRegistry() {
         return kafkaConsumerMetrics;
     }
 
+    public AsyncConsumerMetrics asyncConsumerMetrics() {
```
I don't think this needs to be public. The test is in the same package. For `kafkaConsumerMetrics()`, there is a test which accesses the metrics object from the external object via the delegate.

AndrewJSchofield reviewed (changes requested):
Thanks for the PR.

A couple of other questions.

Could you please check that the share consumer metrics are properly tidied up when the share consumer close?

Also, I'm not sure why `ShareConsumerMetrics` and `KafkaShareConsumerMetrics` both has `CONSUMER_SHARE_METRIC_GROUP_PREFIX` as constructor arguments. I suspect this is just the remains of a previous class hierarchy. I expect this area is ripe for a bit of refactoring too. Seems sensible to do it at the same time.

ShivsundarR committed:
Address review comments

ShivsundarR committed:
Remove redundant argument

ShivsundarR commented:
Hi @AndrewJSchofield , Thanks for the review. I have addressed the comments and updated the PR.

> Could you please check that the share consumer metrics are properly tidied up when the share consumer close?

- Yes I checked this, we are closing all the associated metrics when we close the share consumer.


> I'm not sure why ShareConsumerMetrics and KafkaShareConsumerMetrics both has CONSUMER_SHARE_METRIC_GROUP_PREFIX as constructor arguments

- Yes, ideally we need not have the prefix passed as an argument, but I think to maintain consistency with `KafkaConsumerMetrics` and `ConsumerMetrics`, we might have done this. 

- I have removed the parameter from `KafkaShareConsumerMetrics`, `ShareConsumerMetrics` and `KafkaConsumerMetrics`, `ConsumerMetrics`.


Additionally, we were having the same group name (`consumer-metrics`) for the `AsyncConsumerMetrics` initialised by both `ShareConsumerImpl` and `AsyncKafkaConsumer`.
I have now added a prefix as an argument for `AsyncConsumerMetrics` class.

ShivsundarR committed:
Spotless fix

ShivsundarR committed:
Merge remote-tracking branch 'upstream/trunk' into KAFKA-18220

ShivsundarR committed:
Merge remote-tracking branch 'upstream/trunk' into KAFKA-18220

AndrewJSchofield reviewed (changes requested):
I've done some more testing in advance of merging and I can see that there are some subtle changes in the metrics caused by this PR. I used commands like this to see the metrics and then compared the lists before and after:

```
bin/kafka-consumer-perf-test.sh --bootstrap-server localhost:9092 --topic T1 --print-metrics --messages 1000
```

Similar for share consumer. Please can you take a look.

ShivsundarR commented:
Thanks @AndrewJSchofield, I used the `consumer-perf-test.sh` and `share-consumer-perf-test.sh` tools to print the metrics before and after the changes.

- For `ClassicKafkaConsumer` - There are no changes in the names/number of metrics.
- For `AsyncKafkaConsumer`(group.protocol=consumer) - There are no changes in the names/number of metrics.

- For `ShareConsumerImpl`, there is a difference of 6 metrics. These 6 metrics were appearing additionally before the changes (in trunk) which ideally should not be recorded for share-consumer. With the PR, these 6 metrics no longer appear.
```
These 4 metrics are already recorded as part of KafkaShareConsumerMetrics, previously as AsyncConsumerMetrics inherited KafkaConsumerMetrics, we were getting duplicate metrics.
consumer-metrics:last-poll-seconds-ago
consumer-metrics:poll-idle-ratio-avg
consumer-metrics:time-between-poll-avg
consumer-metrics:time-between-poll-max

These 2 metrics were also part of KafkaConsumerMetrics, but we do not need them for ShareConsumerImpl. These were present before as AsyncConsumerMetrics inherited KafkaConsumerMetrics.
consumer-metrics:commit-sync-time-ns-total
consumer-metrics:committed-time-ns-total
```

Hence the difference is expected.
We have also changes prefix name to include "share" for the `AsyncConsumerMetrics`. This changes prefix for a few metrics but the count remains the same.

chia7712 commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumerImpl.java:
```diff
@@ -1116,19 +1116,23 @@ private boolean processBackgroundEvents() {
 
         LinkedList<BackgroundEvent> events = new LinkedList<>();
         backgroundEventQueue.drainTo(events);
+        if (!events.isEmpty()) {
```
`BackgroundEventHandler#drainEvents` is not used, so the metrics `background-event-queue-size` is never reset to zero. One option could be to call `asyncConsumerMetrics.recordBackgroundEventQueueSize(0)` manually. Alternatively, `ShareConsumerImpl` could use `BackgroundEventHandler` instead of `backgroundEventQueue`, similar to `AsyncKafkaConsumer`

ShivsundarR replied:
Thanks @chia7712 , that makes sense. I have created another PR - https://github.com/apache/kafka/pull/20474 to fix the same, I have used `BackgroundEventHandler::drainEvents` to keep it consistent with `AsyncKafkaConsumer`.
